File: apps/android/feature/scribble/data/src/main/java/com/scribblefit/feature/scribble/data/Mappers.kt
package com.scribblefit.feature.scribble.data

import com.scribblefit.core.database.entity.EntitySyncStatus
import com.scribblefit.core.database.entity.ExerciseEntity
import com.scribblefit.core.database.entity.ScribbleEntity
import com.scribblefit.core.database.entity.SetEntity
import com.scribblefit.core.database.entity.WorkoutEntity
import com.scribblefit.feature.workout.domain.Exercise
import com.scribblefit.feature.workout.domain.Set
import com.scribblefit.feature.workout.domain.Workout
import com.scribblefit.feature.scribble.domain.Scribble
import com.scribblefit.feature.scribble.domain.SyncStatus
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid


internal fun EntitySyncStatus.toDomain(): SyncStatus = when (this) {
    EntitySyncStatus.PENDING -> SyncStatus.Pending
    EntitySyncStatus.FAILED -> SyncStatus.Failed
    EntitySyncStatus.COMPLETED -> SyncStatus.Logged
}

internal fun SyncStatus.toEntity(): EntitySyncStatus = when (this) {
    SyncStatus.Pending -> EntitySyncStatus.PENDING
    SyncStatus.Failed -> EntitySyncStatus.FAILED
    SyncStatus.Logged -> EntitySyncStatus.COMPLETED
    // A completed status in the domain means that there is exercise data available,
    // so we can mark it as completed in the entity.
    // We will fetch the exercise data separately and update the scribble
    // with the exercise ID to ensure data integrity.
    is SyncStatus.Completed -> EntitySyncStatus.COMPLETED
}

internal fun ExerciseEntity.toDomain(sets: List<Set>) = Exercise(
    canonicalName = canonicalName,
    muscleGroup = muscleGroup,
    sets = sets
)

internal fun SetEntity.toDomain() = Set(
    weight = weight,
    reps = reps,
    rpe = rpe,
    notes = notes
)

internal fun ScribbleEntity.toDomain() = Scribble.Raw(
    id = id,
    createdAt = createdAt,
    rawText = rawText,
    status = status.toDomain()
)

@OptIn(ExperimentalUuidApi::class)
internal fun Exercise.toEntity() = ExerciseEntity(
    id = Uuid.random().toString(),
    canonicalName = canonicalName,
    muscleGroup = muscleGroup
)

@OptIn(ExperimentalUuidApi::class)
internal fun Set.toEntity(
    exerciseId: String,
    workoutId: String
) = SetEntity(
    id = Uuid.random().toString(),
    exerciseId = exerciseId,
    workoutId = workoutId,
    weight = weight,
    reps = reps,
    rpe = rpe,
    notes = notes
)

@OptIn(ExperimentalUuidApi::class)
internal fun Workout.toEntity() = WorkoutEntity(
    id = Uuid.random().toString(),
    date = date,
    totalVolume = exercises.sumOf { exercise ->
        exercise.sets.sumOf { set ->
            set.weight * set.reps
        }
    }
)

File: apps/ios/ScribbleFit/Sources/Features/Sync/Data/Mappers.swift
import Foundation

extension ScribbleEntity {
    func toScribble() -> Scribble {
        switch syncStatus {
        case .completed:
            if let data = jsonData?.data(using: .utf8),
               let workout = try? JSONDecoder().decode(ParsedWorkout.self, from: data),
               let firstExercise = workout.exercises.first {
                let syncExercise = SyncExercise(
                    canonicalName: firstExercise.canonicalName,
                    muscleGroup: firstExercise.muscleGroup,
                    sets: firstExercise.sets.map { s in
                        SyncExerciseSet(weight: s.weight, reps: s.reps, rpe: s.rpe, notes: s.notes)
                    }
                )
                return .parsed(id: id, createdAt: createdAt, value: syncExercise)
            }
            return .raw(id: id, createdAt: createdAt, rawText: rawText ?? "", status: .pending)
        case .failed:
            return .raw(id: id, createdAt: createdAt, rawText: rawText ?? "", status: .failed)
        case .pending, .processing:
            return .raw(id: id, createdAt: createdAt, rawText: rawText ?? "", status: .pending)
        }
    }
}
